pkg/utils/kubernetes/certificatesigningrequest/certificate_test.go

```diff
@@ -0,0 +1,138 @@
+			go func(ctx context.Context) {
+				defer GinkgoRecover()
+				Eventually(func(g Gomega) {
+					timeNow := metav1.Now()
+					condition := certificatesv1.CertificateSigningRequestCondition{
+						Type:               certificatesv1.CertificateApproved,
+						Status:             corev1.ConditionTrue,
+						Reason:             "RequestApproved",
+						LastTransitionTime: timeNow,
+						LastUpdateTime:     timeNow,
+					}
+					handleCSR(ctx, g, clientSet, csrPrefix, condition, expectedCertData)
+				}).Should(Succeed())
+			}(ctx)
+
+			certData, privateKeyData, csrName, err := RequestCertificate(ctx, log, clientSet, certificateSubject, dnsSANs, ipSANs, validityDuration, csrPrefix)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(certData).To(Equal(expectedCertData))
+			Expect(privateKeyData).ToNot(BeEmpty())
+			Expect(strings.HasPrefix(csrName, csrPrefix)).To(BeTrue())
+		}, NodeTimeout(time.Second*5))
+
+		It("should return an error if the CSR was denied", func(ctx context.Context) {
+			go func(ctx context.Context) {
+				defer GinkgoRecover()
+				Eventually(func(g Gomega) {
+					timeNow := metav1.Now()
+					condition := certificatesv1.CertificateSigningRequestCondition{
+						Type:               certificatesv1.CertificateDenied,
+						Status:             corev1.ConditionTrue,
+						Reason:             "RequestDenied",
+						LastTransitionTime: timeNow,
+						LastUpdateTime:     timeNow,
+					}
+					handleCSR(ctx, g, clientSet, csrPrefix, condition, nil)
+				}).Should(Succeed())
+			}(ctx)
+
+			_, _, _, err := RequestCertificate(ctx, log, clientSet, certificateSubject, dnsSANs, ipSANs, validityDuration, csrPrefix)
+			Expect(err).To(MatchError(ContainSubstring("certificate signing request is denied")))
+		}, NodeTimeout(time.Second*5))
+
+		It("should return an error if the CSR failed", func(ctx context.Context) {
+			go func(ctx context.Context) {
+				defer GinkgoRecover()
+				Eventually(func(g Gomega) {
+					timeNow := metav1.Now()
+					condition := certificatesv1.CertificateSigningRequestCondition{
+						Type:               certificatesv1.CertificateFailed,
+						Status:             corev1.ConditionTrue,
+						Reason:             "RequestFailed",
+						LastTransitionTime: timeNow,
+						LastUpdateTime:     timeNow,
+					}
+					handleCSR(ctx, g, clientSet, csrPrefix, condition, nil)
+				}).Should(Succeed())
+			}(ctx)
+
+			_, _, _, err := RequestCertificate(ctx, log, clientSet, certificateSubject, dnsSANs, ipSANs, validityDuration, csrPrefix)
+			Expect(err).To(MatchError(ContainSubstring("certificate signing request failed")))
+		}, NodeTimeout(time.Second*5))
+	})
+})
+
+func handleCSR(ctx context.Context, g Gomega, clientSet kubernetesclientset.Interface, csrPrefix string, condition certificatesv1.CertificateSigningRequestCondition, certData []byte) {
+	csrList, err := clientSet.CertificatesV1().CertificateSigningRequests().List(ctx, metav1.ListOptions{})
+	g.Expect(err).NotTo(HaveOccurred())
```

Comment:
The output in case of errors looks equal in both cases.
```
• [TIMEDOUT] [5.003 seconds]
Certificate #RequestCertificate [It] should return the issued certificate
/Users/oliver/SAPDevelop/github.com/gardener/gardener/pkg/utils/kubernetes/certificatesigningrequest/certificate_test.go:57

  [TIMEDOUT] A node timeout occurred
  In [It] at: /Users/oliver/SAPDevelop/github.com/gardener/gardener/pkg/utils/kubernetes/certificatesigningrequest/certificate_test.go:57 @ 11/20/24 14:28:06.535

  This is the Progress Report generated when the node timeout occurred:
    Certificate #RequestCertificate should return the issued certificate (Spec Runtime: 5.001s)
      /Users/oliver/SAPDevelop/github.com/gardener/gardener/pkg/utils/kubernetes/certificatesigningrequest/certificate_test.go:57
      In [It] (Node Runtime: 5s)
        /Users/oliver/SAPDevelop/github.com/gardener/gardener/pkg/utils/kubernetes/certificatesigningrequest/certificate_test.go:57

      Spec Goroutine
      goroutine 39 [select]
        github.com/gardener/gardener/pkg/utils/retry.UntilFor.func1(0x1067eaf38?, {0x1067eaf38, 0x14000166930}, {0x1067dda90, 0x1400004e220})
          /Users/oliver/SAPDevelop/github.com/gardener/gardener/pkg/utils/retry/retry.go:151
        github.com/gardener/gardener/pkg/utils/retry.UntilFor({0x1067eaf38, 0x14000166930}, 0x14000266c80, {0x1067dda90, 0x1400004e220}, 0x14000137d80)
          /Users/oliver/SAPDevelop/github.com/gardener/gardener/pkg/utils/retry/retry.go:162
        github.com/gardener/gardener/pkg/utils/retry.(*ops).Until(0x14000130600, {0x1067eaf38, 0x14000166930}, 0x106692800?, 0x14000137d80)
          /Users/oliver/SAPDevelop/github.com/gardener/gardener/pkg/utils/retry/retry.go:176
      > github.com/gardener/gardener/pkg/utils/kubernetes/certificatesigningrequest.waitForCertificate({0x1067eaf38, 0x14000166930}, {0x1067ff130, 0x1400008c0b0}, {0x140004a06c0, 0x33}, {0x0, 0x0})
          /Users/oliver/SAPDevelop/github.com/gardener/gardener/pkg/utils/kubernetes/certificatesigningrequest/certificate.go:185
            | var certData []byte
            | 
            > if err := retry.Until(ctx, 2*time.Second, func(ctx context.Context) (done bool, err error) {
            | 	if csr, err := client.CertificatesV1().CertificateSigningRequests().Get(ctx, reqName, metav1.GetOptions{}); err == nil {
            | 		if csr.UID != reqUID {
      > github.com/gardener/gardener/pkg/utils/kubernetes/certificatesigningrequest.requestCertificate({0x107f3d838, 0x14000131ce0}, {{0x0?, 0x0?}, 0x4?}, {0x1067ff130, 0x1400008c0b0}, {0x140001d6b00?, 0x107c24a68?, 0x20?}, ...)
          /Users/oliver/SAPDevelop/github.com/gardener/gardener/pkg/utils/kubernetes/certificatesigningrequest/certificate.go:169
            | log.Info("Waiting for certificate signing request to be approved and contain the client certificate")
            | 
            > certData, err = waitForCertificate(childCtx, client, reqName, reqUID)
            | if err != nil {
            | 	return nil, "", err
      > github.com/gardener/gardener/pkg/utils/kubernetes/certificatesigningrequest.RequestCertificate({0x107f3d838, 0x14000131ce0}, {{0x0?, 0x0?}, 0x5?}, {0x1067ff130, 0x1400008c0b0}, 0x140001a2500, {0x1400004e170, 0x1, ...}, ...)
          /Users/oliver/SAPDevelop/github.com/gardener/gardener/pkg/utils/kubernetes/certificatesigningrequest/certificate.go:58
            | }
            | 
            > certData, csrName, err := requestCertificate(ctx, log, client, privateKeyData, certificateSubject, dnsSANs, ipSANs, validityDuration, csrPrefix)
            | if err != nil {
            | 	return nil, nil, "", err
      > github.com/gardener/gardener/pkg/utils/kubernetes/certificatesigningrequest_test.init.func1.2.1({0x107f3d838, 0x14000131ce0})
          /Users/oliver/SAPDevelop/github.com/gardener/gardener/pkg/utils/kubernetes/certificatesigningrequest/certificate_test.go:73
            | }(ctx)
            | 
            > certData, privateKeyData, csrName, err := RequestCertificate(ctx, log, clientSet, certificateSubject, dnsSANs, ipSANs, validityDuration, csrPrefix)
            | Expect(err).NotTo(HaveOccurred())
            | Expect(certData).To(Equal(expectedCertData))
        github.com/onsi/ginkgo/v2/internal.extractBodyFunction.func2({0x1067eeaa8?, 0x14000131ce0?})
          /Users/oliver/SAPDevelop/go/pkg/mod/github.com/onsi/ginkgo/v2@v2.21.0/internal/node.go:465
        github.com/onsi/ginkgo/v2/internal.(*Suite).runNode.func3()
          /Users/oliver/SAPDevelop/go/pkg/mod/github.com/onsi/ginkgo/v2@v2.21.0/internal/suite.go:894
        github.com/onsi/ginkgo/v2/internal.(*Suite).runNode in goroutine 8
          /Users/oliver/SAPDevelop/go/pkg/mod/github.com/onsi/ginkgo/v2@v2.21.0/internal/suite.go:881

  [FAILED] A node timeout occurred and then the following failure was recorded in the timedout node before it exited:
  Timed out after 1.000s.
  The function passed to Eventually failed at /Users/oliver/SAPDevelop/github.com/gardener/gardener/pkg/utils/kubernetes/certificatesigningrequest/certificate_test.go:69 with:
  Expected
      <[]v1.KeyUsage | len:3, cap:3>: ["digital signature", "key encipherment", "client auth"]
  to equal
      <[]v1.KeyUsage | len:4, cap:4>: ["oerks", "digital signature", "key encipherment", "client auth"]
  In [It] at: /Users/oliver/SAPDevelop/github.com/gardener/gardener/pkg/utils/kubernetes/certificatesigningrequest/certificate_test.go:70 @ 11/20/24 14:28:06.536
```